File: python/day22.py
import unittest
from utilities import read_blocks
from collections import defaultdict, deque
from typing import List, Dict, Set, Optional


class Player:
    def __init__(self, deck: List[int]):
        self.deck = deque(deck)

    def is_out_of_card(self):
        return len(self.deck) == 0

    def draw(self) -> Optional[int]:
        if self.is_out_of_card():
            return None
        return self.deck.popleft()

    def push(self, card: int):
        self.deck.append(card)

    def combat(self, other) -> bool:
        while not self.is_out_of_card() and not other.is_out_of_card():
            my_card = self.draw()
            other_card = other.draw()
            if my_card > other_card:
                self.push(my_card)
                self.push(other_card)
            else:
                other.push(other_card)
                other.push(my_card)

    @property
    def score(self) -> int:
        return sum([(i+1) * card for i, card in enumerate(reversed(self.deck))])

    def is_ready_for_subgame(self, num) -> bool:
        return len(self.deck) >= num

    def prepare_subgame(self, num: int):
        return Player(list(self.deck)[:num])

    def __str__(self) -> str:
        return ','.join([str(card) for card in self.deck])

    def recursive_combat(self, other) -> bool:
        memory = set()
        while not self.is_out_of_card() and not other.is_out_of_card():
            s1 = str(self)
            s2 = str(other)
            if (s1, s2) in memory:
                return True
            memory.add((s1, s2))
            my_card = self.draw()
            other_card = other.draw()
            if self.is_ready_for_subgame(my_card) and other.is_ready_for_subgame(other_card):
                sub_self = self.prepare_subgame(my_card)
                sub_other = other.prepare_subgame(other_card)
                if sub_self.recursive_combat(sub_other):
                    self.push(my_card)
                    self.push(other_card)
                else:
                    other.push(other_card)
                    other.push(my_card)
            elif my_card > other_card:
                self.push(my_card)
                self.push(other_card)
            else:
                other.push(other_card)
                other.push(my_card)
        return other.is_out_of_card()


def solve_part_1(players: List[str]) -> int:
    player1 = Player([int(card) for card in players[0].splitlines()[1:]])
    player2 = Player([int(card) for card in players[1].splitlines()[1:]])
    player1.combat(player2)
    return player1.score + player2.score


def solve_part_2(players: List[str]) -> int:
    player1 = Player([int(card) for card in players[0].splitlines()[1:]])
    player2 = Player([int(card) for card in players[1].splitlines()[1:]])
    if player1.recursive_combat(player2):
        return player1.score

    return player2.score


class Testing(unittest.TestCase):

    def test_solution(self):
        test_input = """Player 1:
9
2
6
3
1

Player 2:
5
8
4
7
10"""

        self.assertEqual(solve_part_1(test_input.split('\n\n')), 306)
        self.assertEqual(solve_part_2(test_input.split('\n\n')), 291)


if __name__ == '__main__':
    print('Day 22')
    import time
    start = time.time()
    players: List[str] = read_blocks('../inputs/day22.txt')

    # solution = Solution(foods)
    print('\tPart 1: {}'.format(solve_part_1(players)))
    print('\tPart 2: {}'.format(solve_part_2(players)))

    print("\tTime: {:.4f}ms".format((time.time() - start) * 1000))
    unittest.main()

File: rust/src/day22.rs
use std::collections::hash_map::DefaultHasher;
use std::collections::{HashSet, VecDeque};
use std::hash::{Hash, Hasher};

use crate::utilities::read_block;

#[derive(Debug, Clone)]
struct Player {
    deck: VecDeque<usize>,
}

impl Player {
    fn new(player: &String) -> Self {
        Self {
            deck: player
                .lines()
                .skip(1)
                .map(|card| card.parse().unwrap())
                .collect(),
        }
    }

    fn combat(&mut self, other: &mut Player) -> bool {
        while !(self.is_lost() || other.is_lost()) {
            let self_card = self.draw().unwrap();
            let other_card = other.draw().unwrap();
            let self_win = self_card > other_card;
            self.collect_cards(other, self_win, self_card, other_card);
        }
        other.is_lost()
    }

    fn recursive_combat(&mut self, other: &mut Player) -> bool {
        let mut memory = HashSet::new();
        while !(self.is_lost() || other.is_lost()) {
            let mut hasher = DefaultHasher::new();

            (&self.deck, &other.deck).hash(&mut hasher);
            let game_state = hasher.finish();

            if !memory.insert(game_state) {
                // Found the same game state
                return true;
            }
            let self_card = self.draw().unwrap();
            let other_card = other.draw().unwrap();

            let self_win = if self.can_play_subgame(self_card) && other.can_play_subgame(other_card)
            {
                let mut sub_self = self.prepare_for_subgame(self_card);
                let mut sub_other = other.prepare_for_subgame(other_card);
                sub_self.recursive_combat(&mut sub_other)
            } else {
                self_card > other_card
            };
            self.collect_cards(other, self_win, self_card, other_card);
        }
        other.is_lost()
    }

    fn collect_cards(
        &mut self,
        other: &mut Player,
        self_win: bool,
        self_card: usize,
        other_card: usize,
    ) {
        if self_win {
            self.collect(self_card);
            self.collect(other_card);
        } else {
            other.collect(other_card);
            other.collect(self_card);
        }
    }

    fn can_play_subgame(&self, num: usize) -> bool {
        self.deck.len() >= num
    }

    fn prepare_for_subgame(&self, num: usize) -> Player {
        Player {
            deck: self.deck.iter().take(num).cloned().collect(),
        }
    }

    fn is_lost(&self) -> bool {
        self.deck.is_empty()
    }

    fn draw(&mut self) -> Option<usize> {
        self.deck.pop_front()
    }

    fn collect(&mut self, card: usize) {
        self.deck.push_back(card);
    }

    fn score(&self) -> usize {
        self.deck
            .iter()
            .rev()
            .enumerate()
            .map(|(i, card)| *card * (i + 1))
            .sum()
    }
}

fn solve_part_1(players: &Vec<String>) -> usize {
    let mut player1 = Player::new(&players[0]);
    let mut player2 = Player::new(&players[1]);
    if player1.combat(&mut player2) {
        player1.score()
    } else {
        player2.score()
    }
}

fn solve_part_2(players: &Vec<String>) -> usize {
    let mut player1 = Player::new(&players[0]);
    let mut player2 = Player::new(&players[1]);
    if player1.recursive_combat(&mut player2) {
        player1.score()
    } else {
        player2.score()
    }
}

pub fn main() {
    println!("Day 22");
    let players = read_block("../inputs/day22.txt");
    println!("\tPart1: {}", solve_part_1(&players));
    println!("\tPart2: {}", solve_part_2(&players));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_input = r#"Player 1:
9
2
6
3
1

Player 2:
5
8
4
7
10"#;
        let players: Vec<String> = test_input.split("\n\n").map(|s| s.to_string()).collect();
        assert_eq!(solve_part_1(&players), 306);
        assert_eq!(solve_part_2(&players), 291);
    }
}
